dbms/src/DataStreams/CheckConstraintsBlockOutputStream.cpp

```diff
@@ -0,0 +1,43 @@
+#include <DataStreams/CheckConstraintsBlockOutputStream.h>
+
+
+namespace DB
+{
+
+void CheckConstraintsBlockOutputStream::write(const Block & block)
+{
+    for (auto & constraint_expr: expressions)
+        if (!checkConstraintOnBlock(block, constraint_expr))
+            throw Exception("Some constraints are not satisfied", ErrorCodes::QUERY_WAS_CANCELLED);
+    output->write(block);
+}
+
+void CheckConstraintsBlockOutputStream::flush()
+{
+    output->flush();
+}
+
+void CheckConstraintsBlockOutputStream::writePrefix()
+{
+    output->writePrefix();
+}
+
+void CheckConstraintsBlockOutputStream::writeSuffix()
+{
+    output->writeSuffix();
+}
+
+bool CheckConstraintsBlockOutputStream::checkConstraintOnBlock(const Block & block, const ExpressionActionsPtr & constraint)
+{
+    Block res = block;
+    constraint->execute(res);
+    assert(block.columns() == res.columns() - 1);
+    ColumnWithTypeAndName res_column = res.safeGetByPosition(res.columns() - 1);
+    size_t column_size = res_column.column->size();
+    for (size_t i = 0; i < column_size; ++i)
```

Comment:
A loop with virtual calls.
Better to cast to ColumnUInt8 (and ensure that it is the ColumnUInt8)
and implement quick SIMD check like `memoryIsZero` but another.

Reply:
You mean to cast the result column to `ColumnUInt8` and what to do next? Didn't get how to implement SIMD check here and what to use instead of `memoryIsZero`?

Reply:
> You mean to cast the result column to ColumnUInt8 and what to do next?

Yes. Then access to the internal memory region and implement SIMD check that column doesn't contain zeros.

Reply:
But how I will understand which constraint failed then? Or we should exclude this information from the exception? Or look for the failed constraint only in case of SIMD check failure?

Reply:
@alexey-milovidov up